marshmallow_dataclass/__init__.py

```diff
@@ -442,14 +442,16 @@ def field_for_schema(
                 values=field_for_schema(arguments[1], base_schema=base_schema),
                 **metadata,
             )
-        elif typing_inspect.is_optional_type(typ):
-            subtyp = next(t for t in arguments if t is not NoneType)  # type: ignore
-            # Treat optional types as types with a None default
-            metadata["default"] = metadata.get("default", None)
-            metadata["missing"] = metadata.get("missing", None)
-            metadata["required"] = False
-            return field_for_schema(subtyp, metadata=metadata, base_schema=base_schema)
         elif typing_inspect.is_union_type(typ):
+            if typing_inspect.is_optional_type(typ):
+                metadata["default"] = metadata.get("default", None)
+                metadata["missing"] = metadata.get("missing", None)
+                metadata["required"] = False
+            subtypes = [t for t in arguments if t is not NoneType]  # type: ignore
```

Comment:
Why do we need a `# type: ignore` here ?

Reply:
thanks, I've removed it.

Reply:
@lovasoa Actually it's needed, otherwise mypy hook fails.

Reply:
So, why is it needed ? What is the type error ?

Reply:
it's got here as part of modification of ```subtyp = next(t for t in arguments if t is not NoneType)  # type: ignore``` which were in place before. Do you want to dig into that ?

Reply:
Found pr that line originally came in: https://github.com/lovasoa/marshmallow_dataclass/pull/39/files#diff-be67bc8d0a9f634f273c1261669f88caR392